### ISSUE TITLE
Add storage access restriction at method level

### ISSUE THREAD

**Issue by lock9** [neo-project/neo#829]
Suggestion:
Add additional storage access restriction at method level.

We can potentially increase TPS by running stateless methods concurrently.

Maybe the method could have 3 execution levels:
Stateless
ReadOnly
ReadWrite

I know about the need for the stateless methods (#822),  but I'm not very confident if we need ReadOnly and ReadWrite. Maybe we just need Stateless and Stateful types.

**Comment by igormcoelho** [comment]
By method, you mean `CALL`, right? Or `APPCALL`?
The only issue in executing things in parallel is that you need to sync execution at some point...

I mean, you execute `ContractA` (or `MethodA`) and now you execute `ContractB` (or `MethodB`). When will you stop waiting?

**Comment by shargon** [comment]
or in an attribute inside transaction?

**Comment by lock9** [comment]
We can leave this stateless processing to be done "before or after" the stateful part is done.
Example:
![Screen Shot 2019-06-16 at 20 55 37](https://user-images.githubusercontent.com/38396062/59571610-f3138d00-907c-11e9-9b98-18f3580cfa44.png)
![Screen Shot 2019-06-16 at 20 56 03](https://user-images.githubusercontent.com/38396062/59571615-fd358b80-907c-11e9-96f1-8027f660ccff.png)

We should allow only stateless methods in pre and post-processing stages. 

What do you mean by attribute @shargon ?

This is not mandatory to NEO 3.0, it can be added later.

**Comment by lock9** [comment]
Closing this in favor or #919

**Comment by igormcoelho** [comment]
I think this is more connected to: https://github.com/neo-project/neo/issues/927

Is it different than 927?
